### PR TITLE
Compiler: fix enum overflow on non-default values

### LINKED ISSUE #7268: Enum that inherit UInt64 do not get right values
The compiler uses 32 bit integer to keep the values assigned to enum members.
If the Enum is declared with UInt64 as in https://github.com/crystal-lang/crystal/blob/master/src/llvm/enums.cr#L4 the user needs to assign values explicitly.

Affected lines https://github.com/crystal-lang/crystal/blob/master/src/compiler/crystal/semantic/top_level_visitor.cr#L683-L686

Extracted from #7262

---

$ crystal --version
Crystal 0.27.0

### PR DESCRIPTION
**Pull request by asterite**
Fixes #7268

There's also #7271, which is slightly different, but I'll fix it in a separate PR (it's harder).

### CODE REVIEW

File: spec/compiler/semantic/enum_spec.cr [hunk 1]
```diff
@@ -462,4 +462,14 @@ describe "Semantic: enum" do
...
   end
+
+  it "gives error on enum overflow" do
+    assert_error %(
+      @[Flags]
+      enum Foo : UInt8
+        #{Array.new(9) { |i| "V#{i + 1}" }.join "\n"}
+      end
+      ),
+      "can't automatically compute value of enum member V9 because it would overflow"
```
Comment (straight-shoota) [inline spec/compiler/semantic/enum_spec.cr]: IMO this makes more sense:

```suggestion
      "value of enum member V9 would overflow the base type UInt8"
```

Comment (asterite) [reply spec/compiler/semantic/enum_spec.cr]: I'll try to do it tomorrow

Comment (asterite) [reply spec/compiler/semantic/enum_spec.cr]: Done, thanks for the suggestion!

**Comment by asterite** [comment]
> I guess you are thinking in interpret the constants based on the enum target type

Yes, that's what we are doing right now. So if you have `1 << 40`, it will work fine if the enum is defined as `UInt64` but it will give an error if it's `Int32`. I think that makes sense, there's no point to have the user type `1_u64 << 40` when they expect to get a `UInt64`.

I thought to fix #7271 I just had to catch `OverflowException` but that doesn't seem to be the case.

**Pull request merged**